Adds some high-level documentation to the contracts folder. (#312)
nori-dot-eco/contracts: contracts/README.md

### BridgedPolygonNORI (NORI)

[BridgedPolygonNORI](./BridgedPolygonNORI.sol) is the [$NORI](./NORI.sol) token bridged to Polygon PoS.

- Polygon child chain mapper is the only contract with mint / burn permission.
- Initially deployed with totalSupply of zero.

[NORI on PolygonScan](https://polygonscan.com/token/0x8cf6e82919f69ae382def8f94e581a43ce1e70c1) Currently live deployment (will get replaced with a new version of the contract having ERC777 support removed.)

## Nori Marketplace

### Removal (NRT)

[source](./Removal.sol)

ERC1155 token representing the NRTs (carbon removal credits) issued to a supplier.

Each token ID encodes the wallet address of the supplier to whom the NRTs were issued in addition to information about the source and location of the carbon removed. See [RemovalLib](./RemovalIdLib.sol) for details.

The total supply of each Removal token ID represent the estimated quantity of NRTs (tonnes of CO2 removed from the atmosphere) on a specific parcel of land in a specific year.

Lifecycle of a Removal:

1. Minted to a supplier's address

> 2. Ownership transfered to the _Market_ contract to be listed for sale.

Reviewer: ```suggestion
2. Ownership transferred to the _Market_ contract to be listed for sale.
```

> 3. Sold by the _Market_ and ownership transferred to one or more _Certificate_

Reviewer 2: technically the ownership of the removal is transferred to the Certificate contract in all cases.  the accounting about which certificate IDs "own" the removal exists only in the certificate contract itself. not sure if that's worth being more specific about here

Author: Good point, I was simplifying that but it makes it unclear.

4. Possibly burned or partially burned if the supplier fails to uphold their contractual obligation to keep the underlying carbon sequestered for the duration of their contract.

### Certificate (NCCR)

[source](./Certificate.sol)

ERC721a token representing a carbon removal buyer's proof of purchase.

Every purchase in the market mints a new _Certificate_ token with a unique ID. _Certificate_ is configured as an ERC1155 recipient to receive the purchased _Removal_ tokens.

> Certificates are non-transferrable. It is a fundamental tenet of the Nori market design that Removals and their Certificates cannot change hands after the purchase transaction.

Reviewer: ```suggestion
Certificates are non-transferable. It is a fundamental tenet of the Nori market design that Removals and their Certificates cannot change hands after the purchase transaction.
```

_swap_ mechanism

The _swap_ function is the primary point of interaction with the market for buyers. Calls to the _swap_ function include a quantity of NRTs for purchase and a recipient wallet address to which the Certificate is minted. These calls must additionally include a pre-signed authorization to transfer the corresponding quantity of _BridgedPolygonNORI_ following the [ERC20Permit](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/draft-ERC20Permit.sol) pattern.

The NORI tokens transferred from the buyer to this contract are distributed as follows:

- A configurable market fee percentage to Nori's fee wallet (currently 15%)
- A configurable percentage of the sale proceeds due to the supplier(s) are forwarded to the _RestrictedNORI_ contract to be withheld for insurance purposes and released over the life of their NRT agreement(s). (Configured on each Removal)
- The balance of the proceeds of the sale are forwarded to the supplier(s) address.

## Vesting and Lockup

### LockedNORIV2

[source](./LockedNORIV2.sol)

Investor and Employee timed vesting and unlocking contract.

> Previously audited [here](). The current version was updated to support the new ERC20Permit pattern with _BridgedPolygonNORI_ and to remove reliance on the ERC777 tokensReceived callback.

Reviewer 2: still need a link here I think

[LockedNORI](https://polygonscan.com/token/0xccfffa6c2a030821331cc113b63babdc60bff82a) on PolygonScan.

### RestrictedNORI

Supplier insurance holdback contract.

> Operates similarly to _LockedNORI_ but simplified to remove cliffs and stacked vesting + unlock. Adds the ability to create more than one schedule per owner address and transferability of full or partial restricted balances between addresses.

Reviewer 2: ```suggestion
Operates similarly to _LockedNORI_ by acting as a wrapper token that governs the scheduled release of the underlying _BridgedPolygonNori_ asset.  Implemented as an ERC1155, each token id has its own schedule parameters that control the linear release of the underlying assets over the duration of the schedule.  It is possible to create more than one schedule per owner address and also to transfer full or partial restricted balances between addresses.
```

## Support Libraries

### ArrayLib

### LockedNORILib

The logic for multiple cliffs followed by linear unlock used by the _LockedNORIV2_ contract. The vesting behavior and the unlocking behavior of each grant are each implemented using a schedule from this library.

### RestrictedNORILib

The schedule logic used in _RestrictedNORI_

### RemovalQueue

> The queueing mechanism used by the _Market_ contract to maintain an ordered list of _Removal_ tokens listed for sale on behalf of a given supplier.

Reviewer: ```suggestion
The queuing mechanism used by the _Market_ contract to maintain an ordered list of _Removal_ tokens listed for sale on behalf of a given supplier.
```
